### ISSUE TITLE
Resurfaced: packet header mismatch

### ISSUE THREAD

**Issue by WCPistolPete** [SmartlyDressedGames/Unturned-3.x-Community#2821]
Linux servers were fine yesterday. After a shutdown, update check and restart the "packet header mismatch" error showed up.

May be a repeat of https://github.com/SmartlyDressedGames/Unturned-3.x-Community/issues/2616

JustOneMoreBlock can provide more information later as I don't have direct access to server logs.

**Comment by SDGNelson** [comment]
Where does it say "packet header mismatch"? I do not think there is such an error in Unturned itself. Also nothing has changed in the game files since the Friday update.

**Comment by WCPistolPete** [comment]
Our status page, https://justplayhere.com/status returns the packet header mismatch. The servers are not visible in the in-game browser. unturned-servers.net uses a method to query the servers that is different from the web status page and the in-game browser. The servers show as online to them.
Please refer to the link I posted.

There was a steam update. Maybe it broke steamcmd for linux again. My windows test servers are still visible and players can join those.

Your last statement from the last time this happened: "Everything should be working by default, and I will re-enable the steamcmd copying in a few weeks once the steamcmd version of steamclient.so is patched."

**Comment by CyberAndrii** [comment]
Its a security feature Valve made enabled by default in the last update. You will need to update your site to support it, or you can try disabling it by setting `STEAM_GAMESERVER_A2S_INFO_REQUIRE_CHALLENGE` environment variable to `0`. 

[_Source_](https://steamcommunity.com/discussions/forum/14/2974028351344359625)

**Comment by SDGNelson** [comment]
Thanks @CyberAndrii, that is exciting news!

It looks like the [Linux redist depot was updated yesterday](https://steamdb.info/depot/1006/history/) so maybe your servers just need to be updated and restarted?

**Comment by WCPistolPete** [comment]
Our servers shutdown, check for updates then restart on a two-hour cycle. I really hate the two hour cycle. We lose players to it, but if we don't cycle them, then patches followed by a quick patch to fix a patch-induced bug would keep them offline much longer. They've been kicking this error for a full day. That's 12 update checks and 12 restarts.

Forget about the web-based status page. The servers are not visible to the in-game browser, the same as if they were offline. The last time it was an issue with the steamclient.so file not copying properly.

**Comment by SDGNelson** [comment]
I just spun up a Linux server on ec2 and was able to connect normally, and it shows in the server browser.
IP: 54.184.220.218 Port: 27015

Are there any errors in your server log?

**Comment by WCPistolPete** [comment]
You'd have to ask JustOneMoreBlock about the logs. I don't have access.

**Comment by SDGNelson** [comment]
Will shutdown my server for the meantime. It sounds like this is probably a configuration issue, but @JustOneMoreBlock feel free to comment with any questions or errors.

**Comment by SDGNelson** [comment]
@WCPistolPete clarifying on the questions from you email here so others can see the answers as well:

It looks like justplayhere has a [custom script for querying the server status](https://github.com/JustOneMoreBlock/unturned-php-status) used on the website. If Valve's new anti-spam security enhancements have gone live then the custom scripts would need to be updated. Fletcher goes into detail on how and why in the post linked by SDGNelson.

There are two places the Linux dedicated server can load the most recent steamclient.so redist from. One is the Valve-recommended redist depot which most dedicated servers use. The other is from the steamcmd install folder. Sometimes Valve forgets to update the redist depot with fixes, for example with the dedicated server certificate expiry bug. For that reason it can sometimes be better to use the up-to-date steamcmd version of the file. Currently there redist copy, which the server is using, seems to be fine.

I think it is a server configuration issue because as far as I have heard JPH is the only one affected, and the test Linux server was fine.

**Comment by ModernMoGamer** [comment]
Hello,

I too am receiving issues with querying server data. For my Server Details Discord Bot, I utilize Valve's A2S Query https://developer.valvesoftware.com/wiki/Server_queries

About 4 months ago, I had this issue where server's queried on my Unturned Install's would be unable to read the returned query. This was due to an update to the steamclient.so file, Mixing around different steamclient.so files on my Linux server's solved the issue.

Yesterday I noticed all my Linux installs again failed to reply. I did some light checking with the steamclient.so file's to no avail.
Yesterday I had installed a Windows Unturned Server, no problems responding to the A2S queries.
Today after the recent game update, I still get the issues on Linux which now affect my fresh Windows Unturned Install's.

As CyberAndrii mentioned about Challenges, I looked into my A2S queries, They do not handle challenges.
I tested the demo of https://github.com/Yepoleb/python-a2s on both Windows and Linux server's, neither worked either.

Nelson you mentioned the Redist Depot updates. I believe updating to this new steamclient.so and its windows equivalent steamclient.dll has caused these issues. In the past you've bundled the game with working version inside the Unturned/Linux64 folder that worked, while Steamcmd's steamclient.so didn't. Are you able to provide an older version (atleast to me) so I can test?
I'm not sure where to download older versions listed from https://steamdb.info/depot/1006/manifests/ (64 bit version)

Thanks.

**Comment by SDGNelson** [comment]
@ModernMoGamer 

> As CyberAndrii mentioned about Challenges, I looked into my A2S queries, They do not handle challenges.

If I understand correctly, you are saying your Discord bot does not handle the A2S challenges? If so it sounds like your bot would need to be updated to handle the challenges. Responding to the challenge is necessary to prevent reflection attacks.

**Comment by ModernMoGamer** [comment]
You are correct, my bot does not handle challenges.

Though as I stated, “I tested the demo of https://github.com/Yepoleb/python-a2s on both Windows and Linux server's, neither worked either”.
Yepoleb/python-a2s has sufficient code to handle challenges, though utilizing their demo site, it reports the servers are offline, indicating (afaik) an issue outside of challenges.

**Comment by SDGNelson** [comment]
Sorry I misunderstood. I just checked a few Unturned servers using the demo application you linked and it seemed to work okay. Did you add +1 for the query port?

**Comment by JustOneMoreBlock** [comment]
- Using latest `steamclient.so` from `SteamCMD`.
- The one from the `Unturned_Headless` likely still works.
- No errors in `CONSOLE`.
- This `STEAM_GAMESERVER_A2S_INFO_REQUIRE_CHALLENGE` was no help.
- https://github.com/JustOneMoreBlock/unturned-php-status - Yes, I wrote this. Probably will have to update this to support the new versions. It's now moved to https://status.justplayhere.com/

Note:
@WCPistolPete is going to be sending you an email related to this issue. But, seems a force uninstall/install of Steam corrected the issue. Really annoying.

_Tired of fixing your game. But, we'll blame Steam for this one._

**Comment by WCPistolPete** [comment]
Steam says it updates the client, but sometimes it really doesn't. I could see only 208 out of 2801 servers. I compared my main laptop, which couldn't see 93% of the servers out there, to my backup laptop that could see all of the servers. My main computer updates regularly. The backup... Maybe once every month and a half. The images show the file differences in two supposedly updated computers. The one on the left received two steam updates back on the 13th, but it really didn't update. That was the problem laptop. The files on the backup did update. Steam is only doing a cursory check and not properly updating files.

I had to create a new steam install to get the newest files, then transfer the game files into the new folder. What a pain in the butt. The new steam install, which I shouldn't have needed to do since Steam kept saying it was already updated, fixed the issue of not being able to see the servers. Left side is laptop that Steam supposedly updated. Right side is the backup laptop which actually updated. 

![Versions](https://user-images.githubusercontent.com/53093137/133914089-4aa0607d-d4a3-4b8b-9b01-220753d547b0.jpg)

All is good after the forced steam install.

![Forced Steam Install](https://user-images.githubusercontent.com/53093137/133914156-00645c63-1d3a-4831-a724-15bfae211b48.jpg)

What a royal pain. Hopefully Valve will implement better quality control in the (very near) future.

tldr:
New Steam install may be needed to properly update client files even if Steam says everything is up to date.

**Comment by SDGNelson** [comment]
Glad to hear you got it working!

And @ModernMoGamer I assume it was the +1 port difference?

**Comment by ModernMoGamer** [comment]
> Glad to hear you got it working!
> 
> And @ModernMoGamer I assume it was the +1 port difference?

As there wasn’t a separate port spot, I forgot I had to specify it manually instead of it pulling the Unturned Default. 
Will need to update my services to follow the new captcha.
Thank you for the clarification
